### PR TITLE
Consider `NaN` to be equal to itself in server protocol tests

### PR DESCRIPTION
**Pull request by david-perez**
The client protocol tests use the
`aws_smithy_protocol_test::FloatEquals` for this [0].

Note we're only applying this to direct floating point shape members,
i.e. this commit _does not_ address #1147.

[0]: https://docs.rs/aws-smithy-protocol-test/latest/aws_smithy_protocol_test/trait.FloatEquals.html

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by crisidev — COMMENTED** [review]
Isn't this semantically wrong? https://doc.rust-lang.org/std/primitive.f32.html

**Comment by david-perez** [comment]
@crisidev In protocol tests we don't want to do any math on `NaN` values, nor do we care about the specific bits of a _particular_ `NaN`, or about what calculation poisoned the result. We do need however a mechanism to verify that we deserialize `"NaN"` on the wire to a `NaN` floating point value, i.e. to check for the _presence_ of _a_ (any) `NaN` value. Using `aws_smithy_protocol_test::FloatEquals` for this purpose is fair use.

I found this post interesting https://stackoverflow.com/questions/10034149/why-is-nan-not-equal-to-nan

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt [hunk 1]
```diff
@@ -357,7 +359,40 @@ class ServerProtocolTestGenerator(
...
+                                    #{AssertEq}(parsed.$memberName, expected.$memberName, "Unexpected value for `$memberName`");
+                                    """,
+                                *codegenScope
+                            )
+                        }
+                    }
+                }
+            } else {
+                rustWriter.rustTemplate("#{AssertEq}(parsed, expected);", *codegenScope)
+            }
```
Comment (rcoh) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:395]: should we being extracting some sort of `ProtocolTestCore`? The client has this behavior too right?

**Review by rcoh — APPROVED** [review]

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt (hunk 1 above)
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:395]: 100%. It's slightly different in that the client _always_ iterates over the members to do the equality comparison.

But yes, once I get all the protocol tests for restJson1 passing, I plan on extracting a lot of shared functionality between client and server protocol tests.

**Review by crisidev — APPROVED** [review]

**Pull request merged**